Create Navigation when restoring all items in back-forward list

sysrqb opened the pull request:
#### 2ede4615af7ace55e5aade5800388a61d4c9d9c5
<pre>
Create Navigation when restoring all items in back-forward list
<a href="https://bugs.webkit.org/show_bug.cgi?id=260636">https://bugs.webkit.org/show_bug.cgi?id=260636</a>
rdar://problem/114351057

Reviewed by Chris Dumez.

If we restore from saved session state and the restored state contained a
back/forword list, we won&apos;t create a Navigation for the current item. As a
result, we&apos;ll use 0 as the placeholder navigationID when we set the pending API
request. We eventually create a new API::Navigation in
WebPageProxy::decidePolicyForNavigationAction() but we don&apos;t update
PageLoadState::m_committedState.pendingAPIRequest if that was initalized, and
this creates an inconsistency between the new navigationID and the cached value
in PageLoadState::m_committedState.pendingAPIRequest.navigationID when we
compare them in later steps.

In this patch, I create an API::Navigation regardless of the current item we
are restoring. This patch adds a new API test.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::goToBackForwardItem):
* Tools/TestWebKitAPI/Tests/WebKit/RestoreSessionState.cpp:
(TestWebKitAPI::decidePolicyForNavigationActionIgnore):
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/267579@main">https://commits.webkit.org/267579@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/38217315dafddabad65c2b912ebb0124c06bc04c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16811 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17134 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17588 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18593 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15748 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17001 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20367 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17273 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/18023 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17010 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17377 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14555 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19437 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14622 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15239 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22024 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15610 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15406 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19766 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16011 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13579 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15186 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4073 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19550 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15842 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -2083,10 +2080,10 @@ RefPtr<API::Navigation> WebPageProxy::goToBackForwardItem(WebBackForwardListItem
     m_process->startResponsivenessTimer();
 
-    return navigation;
+    return navigation.ptr();
```
I am confident we can avoid the ref-counting churn by doing something like this (or similar):
```
return RefPtr<API::Navigation> { WTFMove(navigation) };
```

cdumez reviewed (approved):
LGTM with nit.

sysrqb committed:
Create Navigation when restoring all items in back-forward list
https://bugs.webkit.org/show_bug.cgi?id=260636
rdar://problem/114351057

Reviewed by Chris Dumez.

If we restore from saved session state and the restored state contained a
back/forword list, we won't create a Navigation for the current item. As a
result, we'll use 0 as the placeholder navigationID when we set the pending API
request. We eventually create a new API::Navigation in
WebPageProxy::decidePolicyForNavigationAction() but we don't update
PageLoadState::m_committedState.pendingAPIRequest if that was initalized, and
this creates an inconsistency between the new navigationID and the cached value
in PageLoadState::m_committedState.pendingAPIRequest.navigationID when we
compare them in later steps.

In this patch, I create an API::Navigation regardless of the current item we
are restoring. This patch adds a new API test.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::goToBackForwardItem):
* Tools/TestWebKitAPI/Tests/WebKit/RestoreSessionState.cpp:
(TestWebKitAPI::decidePolicyForNavigationActionIgnore):
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/267579@main
